spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala

```diff
@@ -73,6 +73,44 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
    */
   var gradInput: A = Activity.allocate[A, T]()
 
+
+  private[bigdl] var inputShapeValue: Activity = null
+
+  private[bigdl] var outputShapeValue: Activity = null
+
+  /**
+   * There's no batch dim in the inputShape which just represent a sample record.
+   */
+  def getInputShape(): Activity = {
+    inputShapeValue
+  }
+
+  /**
+   * There's no batch dim in the outputShape which just represent a sample record.
+   */
+  def getOutputShape(): Activity = {
+    if (outputShapeValue == null) {
+      throw new RuntimeException("You should build this model first before getting the outputshape")
+    }
+    outputShapeValue
+  }
+
+  def isBuilt(): Boolean = {
+    getOutputShape() != null
+  }
+
+  private[bigdl] def build(inputShape: Activity): Unit = {
```

Comment:
should we put this logic into container?

Reply:
For a given layer `compile` can tell us what's it's `inputshape`  or inferring information.
and `build` can be use to enrich the `outputshape` or any other meta info need to be inferred..

Reply:
can you give an example? And from the name, I really can't find out the difference...